### PR TITLE
[FLINK-28178][runtime-web] Show the delegated StateBackend and whethe…

### PR DESCRIPTION
**Pull request by zoltar9264**
…r changelog is enabled in the UI

## What is the purpose of the change

Show the delegated StateBackend and whether changelog is enabled in the UI, described in [FLINK-28178](https://issues.apache.org/jira/browse/FLINK-28178).

## Brief change log

  - set field 'state_backend' in /jobs/:jobid/checkpoints/config to delegated StateBackend
  - /jobs/:jobid/checkpoints/config add field 'state_changelog', which indicates whether the State Changelog enabled

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by fredia** [comment]
Thanks for creating this PR @zoltar9264, It looks like you added `state_changelog` to rest api, could you please update the rest api doc?  See [rest-api-documentation](https://github.com/apache/flink/tree/master/flink-docs#rest-api-documentation) for details.

**Comment by zoltar9264** [comment]
Thanks a lot for your reminding @fredia ! I have push a commit fix it.

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java [hunk 1]
```diff
@@ -176,6 +176,13 @@ public interface AccessExecutionGraph extends JobStatusProvider {
      */
     Optional<String> getCheckpointStorageName();
 
+    /**
+     * Returns whether the state changelog is enabled for this ExecutionGraph.
+     *
+     * @return true, if state changelog enabled, false otherwise.
+     */
+    boolean isStateChangelogEnabled();
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:184]: I think `isChangelogStateBackendEnabled` looks better.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java [hunk 2]
```diff
@@ -65,6 +65,8 @@ public class CheckpointConfigInfo implements ResponseBody {
     public static final String FIELD_NAME_CHECKPOINTS_AFTER_TASKS_FINISH =
             "checkpoints_after_tasks_finish";
 
+    public static final String FIELD_NAME_STATE_CHANGELOG = "state_changelog";
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java:68]: I think `changelog_state_enabled` looks better.

File: flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html [hunk 3]
```diff
@@ -551,15 +551,15 @@
             </td>
           </tr>
           <tr>
-            <td>Changelog state-backend</td>
-            <td *ngIf="checkPointConfig['state_backend'] === 'ChangelogStateBackend'">Enabled</td>
-            <td *ngIf="checkPointConfig['state_backend'] !== 'ChangelogStateBackend'">Disabled</td>
+            <td>State Changelog</td>
```
Comment (Myasuka) [inline flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html:554]:
```suggestion
            <td>Changelog State</td>
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java [hunk 4]
```diff
@@ -391,6 +400,7 @@ public static ArchivedExecutionGraph createSparseArchivedExecutionGraph(
                 checkpointingSettings == null ? null : CheckpointStatsSnapshot.empty(),
                 checkpointingSettings == null ? null : "Unknown",
                 checkpointingSettings == null ? null : "Unknown",
+                false,
```
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java:403]: This would always be `false`?

**Review by Myasuka — COMMENTED** [review]
Thanks for creating this PR. Please take a look at my comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java (hunk 1 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:184]: Hi @Myasuka , thanks for the suggession .I think ’State Changelog‘ is more suitable to describe this feature, and ’ChangelogStateBackend‘ is the specific implementation of it. WDYT ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java (hunk 2 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java:68]: How about 'state_changelog_enabled' ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java (hunk 4 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraph.java]: Sorry @Myasuka , It's my cursoriness.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java (hunk 1 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:184]: I think isStateChangelogEnabled is easier to understand, but I've also noticed that isChangelogStateBackendEnabled has been used elsewhere. I don't know how to choose, or is my understanding simply wrong?

Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:184]: To me, making it consistent with the other parts of code makes sense (so `isChangelogStateBackendEnabled` seems preferrable); otherwise, I don't have a strong opinion on this.

File: flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html (hunk 3 above)
Comment (rkhachatryan) [reply flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html:554]: I think it should actually be `State Changelog` - i.e. "Changelog of Flink State";
`Changelog State` reads as "state of changelog of something".

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java (hunk 1 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecutionGraph.java:184]: Actually I think all place where `isChangelogStateBackendEnabled` appeared should be `isStateChangelogEnabled`. But since these methods are not public API for users , I can also accept `isChangelogStateBackendEnabled` for consistence.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java (hunk 2 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java]: Hi @Myasuka , I change it to 'state_changelog_enabled'.

File: flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html (hunk 3 above)
Comment (zoltar9264) [reply flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html:554]: Hi @Myasuka , how do you think about “Changelog of Flink State” ?

**Comment by rkhachatryan** [comment]
I was trying to test the changes locally. 
The UI shows commit `8788d51` but no changelog configuration shown. I do see changelog is used in the logs. 
@zoltar9264 could you please verify?

**Comment by zoltar9264** [comment]
Sorry @rkhachatryan , you are right. I test it on yarn cluster before and it works. I test it locally just now, it really does't work. I will check it.

File: flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html [hunk 5]
```diff
@@ -551,15 +551,15 @@
...
+            <td *ngIf="checkPointConfig['state_changelog_enabled']">Enabled</td>
```
[the remark this answers is not in the document]
Comment (zoltar9264) [reply flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.html:555]: Thanks @fredia for your reminder !

**Comment by zoltar9264** [comment]
Kindly ping @Myasuka @rkhachatryan .

**Comment by rkhachatryan** [comment]
Sorry for the delay @zoltar9264 .
I still don't see the changes when running locally (commit SHA shown in the UI: 7259856).

In the logs, I see messages like this:
```
2022-08-03 21:14:19,963 INFO  org.apache.flink.state.changelog.ChangelogKeyedStateBackend  [] - snapshot of SlidingWindowCheckMapper -> Sink: Sl...
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java [hunk 6]
```diff
@@ -517,7 +527,16 @@ public void failJobDueToTaskFailure(
...
+        if (StateBackendLoader.isChangelogStateBackend(checkpointStateBackend)) {
+            this.stateChangelogEnabled = TernaryBoolean.TRUE;
+            StateBackend delegatedStateBackend =
+                    ((DelegatingStateBackend) checkpointStateBackend).getDelegatedStateBackend();
+            this.stateBackendName = delegatedStateBackend.getClass().getSimpleName();
+        } else {
+            this.stateChangelogEnabled = TernaryBoolean.FALSE;
+            this.stateBackendName = checkpointStateBackend.getClass().getSimpleName();
+        }
+
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java:539]: Can we replace `if` with class-dependent logic here?
For example
```
this.stateBackendName = checkpointStateBackend.getName();
this.stateChangelogEnabled =
        TernaryBoolean.fromBoolean(
                StateBackendLoader.isChangelogStateBackend(checkpointStateBackend));
```

Where `getName()` by default returns `this.getClass().getSimpleName()`,
and changelog returns `delegatedStateBackend.getName()`?

File: docs/layouts/shortcodes/generated/rest_v1_dispatcher.html [hunk 7]
```diff
@@ -2335,6 +2335,9 @@
     "state_backend" : {
       "type" : "string"
     },
+    "state_changelog_enabled" : {
+      "type" : "boolean"
+    },
```
Comment (rkhachatryan) [inline docs/layouts/shortcodes/generated/rest_v1_dispatcher.html:2340]: Is this actually related to this PR? Or #20103?

**Review by rkhachatryan — COMMENTED** [review]
Thanks for updating the PR @zoltar9264 ,
The UI changes looks good to me.
I have a minor suggestion around the code, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java (hunk 6 above)
Comment (zoltar9264) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java]: Thanks @rkhachatryan , good idea, I have modified as your suggestion.

File: docs/layouts/shortcodes/generated/rest_v1_dispatcher.html (hunk 7 above)
Comment (zoltar9264) [reply docs/layouts/shortcodes/generated/rest_v1_dispatcher.html:2340]: Hi @rkhachatryan , here is [#20103](https://github.com/apache/flink/pull/20103) .

**Review by rkhachatryan — APPROVED** [review]
Thanks for updating the PR @zoltar9264 ,
LGTM

**Comment by zoltar9264** [comment]
Thanks @rkhachatryan , I have squash all commit into one.

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by zoltar9264** [comment]
@flinkbot run azure

**Comment by rkhachatryan** [comment]
The build [succeeded](https://dev.azure.com/khachatryanroman/flink/_build/results?buildId=1601&view=results) in a private branch, merging.

**Comment by zoltar9264** [comment]
Thanks @rkhachatryan !
